### ISSUE TITLE
Packaging on Musl / Alpine Linux.

### ISSUE THREAD

**Issue by GamePlayer-8** [ilya-zlobintsev/LACT#400]
Hello,

Any chance that this software will be available on Alpine Linux / Musl LibC?

**Comment by ilya-zlobintsev** [comment]
You can already build it from source with musl. There was a fix for it some time ago.

Prebuilt packages for Alpine should be doable, the packaging tool supports it. I'll look into it.

**Comment by andar1an** [comment]
I am trying to build for alpine at the moment, issue I am running into states  "Unable to find libclang: "the `libclang` shared library at /usr/lib/llvm20/lib/libclang.so.20.1.7 could not be opened: Dynamic loading not supported""

I don't know much about C, but installed clang20 and clang20-libclang from alpine apk. I should have the correct library as per: https://pkgs.alpinelinux.org/contents?name=clang20-libclang&repo=main&branch=edge&arch=x86_64

I am wondering if I need to try to figure out setting static feature as build dependency in bindgen? along lines of this https://github.com/rust-lang/rust-bindgen/issues/2360#issuecomment-2367084230

Not really sure atm.

Alpine is a real challenge to control fan speeds on newer amd gpus, I am on latest stable kernel in aports (6.15), and have tried to manipulate bitmasks for @PP_OD_FUZZY_FAN_CONTROL_MASK because PowerPlay will not let me vary pwm1_enable from 2 ->1, and praying LACT can do this.

**Comment by ilya-zlobintsev** [comment]
> I am trying to build for alpine at the moment, issue I am running into states "Unable to find libclang: "the `libclang` shared library at /usr/lib/llvm20/lib/libclang.so.20.1.7 could not be opened: Dynamic loading not supported""
> 
> I don't know much about C, but installed clang20 and clang20-libclang from alpine apk. I should have the correct library as per: https://pkgs.alpinelinux.org/contents?name=clang20-libclang&repo=main&branch=edge&arch=x86_64
> 
> I am wondering if I need to try to figure out setting static feature as build dependency in bindgen? along lines of this [rust-lang/rust-bindgen#2360 (comment)](https://github.com/rust-lang/rust-bindgen/issues/2360#issuecomment-2367084230)
> 
> Not really sure atm.
> 
> Alpine is a real challenge to control fan speeds on newer amd gpus, I am on latest stable kernel in aports (6.15), and have tried to manipulate bitmasks for @PP_OD_FUZZY_FAN_CONTROL_MASK because PowerPlay will not let me vary pwm1_enable from 2 ->1, and praying LACT can do this.

This error happens because rust defaults to statically linking libc when compiling for musl, building with `RUSTFLAGS="-C target-feature=-crt-static"` to link it dynamically (like for glibc) resolves it.

**Comment by andar1an** [comment]
Thank you, I have it building, but I am stuck where I was without LACT. My bitmask doesn't seem to update whether I change the kernel paramaters for bootloader, or add configuration to modprobe. I still remain with bitmask of 0xfff7bfff. On the plus the openrc service didn't require much tweaking at all for alpine. 

Update: It is all working! I had a comma where I shouldn't have in grub config. lol. Thank you! I need to make a note to try to get this in Aports, there is no reason it shouldn't be, that was not so bad.

**Comment by andar1an** [comment]
> > I am trying to build for alpine at the moment, issue I am running into states "Unable to find libclang: "the `libclang` shared library at /usr/lib/llvm20/lib/libclang.so.20.1.7 could not be opened: Dynamic loading not supported""
> > I don't know much about C, but installed clang20 and clang20-libclang from alpine apk. I should have the correct library as per: https://pkgs.alpinelinux.org/contents?name=clang20-libclang&repo=main&branch=edge&arch=x86_64
> > I am wondering if I need to try to figure out setting static feature as build dependency in bindgen? along lines of this [rust-lang/rust-bindgen#2360 (comment)](https://github.com/rust-lang/rust-bindgen/issues/2360#issuecomment-2367084230)
> > Not really sure atm.
> > Alpine is a real challenge to control fan speeds on newer amd gpus, I am on latest stable kernel in aports (6.15), and have tried to manipulate bitmasks for @PP_OD_FUZZY_FAN_CONTROL_MASK because PowerPlay will not let me vary pwm1_enable from 2 ->1, and praying LACT can do this.
> 
> This error happens because rust defaults to statically linking libc when compiling for musl, building with `RUSTFLAGS="-C target-feature=-crt-static"` to link it dynamically (like for glibc) resolves it.

Hi @ilya-zlobintsev, should it possible to link this statically for musl? I would like to try to get this into aports, and if I can avoid dynamic linking that would be great. Not sure why static linking doesn't work yet, but im sure static libraries should exist in alpine already, and I had clang 20 static libs installed.

**Comment by ilya-zlobintsev** [comment]
No, static linking is not possible, as it is fundamentally incompatible with driver libraries being loaded dynamically (libdrm for amd/intel and nvml for nvidia). GTK doesn't support static linking either.

**Comment by andar1an** [comment]
> No, static linking is not possible, as it is fundamentally incompatible with driver libraries being loaded dynamically (libdrm for amd/intel and nvml for nvidia). GTK doesn't support static linking either.

Understood, thank you for the information.

**Comment by andar1an** [comment]
Started to package. Still a lot to do, and need to consider how many packages e.g. should cli and daemon and client be separate, but it is underway. https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/87654/. Will pick away at it when I can

**Comment by andar1an** [comment]
@ilya-zlobintsev if you have a moment, one of the parts that may take the most time is figuring out which of the build depends are runtime depends exclusively. Is there somewhere that I can see that easily? 

If not I think the gentoo build will get me there.

**Comment by ilya-zlobintsev** [comment]
The arch package should be a good reference (except that libadwaita is not required for the default build): https://gitlab.archlinux.org/archlinux/packaging/packages/lact/-/blob/main/PKGBUILD?ref_type=heads#L12

**Comment by ilya-zlobintsev** [comment]
Additionaly, I'd recommend avoiding packaging a separate libadwaita version, as it is planned to be dropped in the future: https://github.com/ilya-zlobintsev/LACT/issues/526

**Comment by andar1an** [comment]
> Additionaly, I'd recommend avoiding packaging a separate libadwaita version, as it is planned to be dropped in the future: [#526](https://github.com/ilya-zlobintsev/LACT/issues/526)

Thanks so much for saying this haha, I was creating a suboackage for it

**Comment by andar1an** [comment]
As an update, I have made a fair bit of progress, but I am just waiting for the next version of musl to be released as it contains a renameat2 wrapper, which will prevent the need to patch easy_fuser. The next musl release seems to be expected expected within the month, and even if it takes a little longer I believe it makes more sense to just wait a little bit instead of require a patch.

**Comment by knuxyl** [comment]
> No, static linking is not possible, as it is fundamentally incompatible with driver libraries being loaded dynamically (libdrm for amd/intel and nvml for nvidia). GTK doesn't support static linking either.

libdrm supports static linking very easily. i don't know if opengl or vulkan calls can be statically linked. people say they can't, but people also say you can't statically link libdrm. what i do know is libdrm has nothing to do with fundamental incompatibilities with static linking.

I just got lact to build successfully with `export RUSTFLAGS="-C target-feature=-crt-static"`. thank you for sharing @ilya-zlobintsev

**Comment by ilya-zlobintsev** [comment]
> > No, static linking is not possible, as it is fundamentally incompatible with driver libraries being loaded dynamically (libdrm for amd/intel and nvml for nvidia). GTK doesn't support static linking either.
> 
> libdrm supports static linking very easily. i don't know if opengl or vulkan calls can be statically linked. people say they can't, but people also say you can't statically link libdrm. what i do know is libdrm has nothing to do with fundamental incompatibilities with static linking.
> 
> I just got lact to build successfully with `export RUSTFLAGS="-C target-feature=-crt-static"`. thank you for sharing [@ilya-zlobintsev](https://github.com/ilya-zlobintsev)

libdrm is statically linkable (you'd need to compile it from source as a static library to do so, but it is indeed doable), but NVML is a complete no go as the library version needs to match the version of the kernel driver on the current system.

**Comment by andar1an** [comment]
I decided to patch easy_fuser for fakeroot on alpine build because I am not sure when next version of musl is being released. 

I am currently excluding integration tests for down stream easy_fuser because tempfile::TempDir is being problematic, and I don't want to dive in. 

Is easy_fuser used for a lot in LACT @ilya-zlobintsev? It doesn't seem so at quick glance, but I am not sure if it is okay to skip the integrations tests. 

The patches for easy fuser on alpine relate to using renameat instead of renameat2

**Comment by ilya-zlobintsev** [comment]
> I decided to patch easy_fuser for fakeroot on alpine build because I am not sure when next version of musl is being released.
> 
> I am currently excluding integration tests for down stream easy_fuser because tempfile::TempDir is being problematic, and I don't want to dive in.
> 
> Is easy_fuser used for a lot in LACT [@ilya-zlobintsev](https://github.com/ilya-zlobintsev)? It doesn't seem so at quick glance, but I am not sure if it is okay to skip the integrations tests.
> 
> The patches for easy fuser on alpine relate to using renameat instead of renameat2

That library is only used for tests, if you're not trying to run tests, it should not even need to be compiled.

**Comment by andar1an** [comment]
I am running the tests, and lact tests successfully run, it is just the easy_fuser integration tests that fail. Not sure if they are supposed to run, but they may be because of how I patched.

**Comment by ilya-zlobintsev** [comment]
I think it'd be acceptable for the package to skip those tests.
